openedx/core/djangoapps/notifications/base_notification.py

```diff
@@ -65,11 +69,99 @@
 }
 
 
+class NotificationPreferenceSyncManager:
+    """
+    Sync Manager for Notification Preferences
+    """
+
+    @staticmethod
+    def normalize_preferences(preferences):
+        """
+        Normalizes preferences to reduce depth of structure.
+        This simplifies matching of preferences reducing effort to get difference.
+        """
+        apps = []
+        prefs = []
+        non_editable = {}
+        core_notifications = {}
+
+        for app, app_pref in preferences.items():
+            apps.append({
+                'name': app,
+                'enabled': app_pref.get('enabled')
+            })
+            for pref_name, pref_values in app_pref.get('notification_types', {}).items():
+                prefs.append({
+                    'name': pref_name,
+                    'app_name': app,
+                    **pref_values
+                })
+            non_editable[app] = app_pref.get('non_editable', {})
+            core_notifications[app] = app_pref.get('core_notification_types', [])
+
+        normalized_preferences = {
+            'apps': apps,
+            'preferences': prefs,
+            'non_editable': non_editable,
+            'core_notifications': core_notifications,
+        }
+        return normalized_preferences
+
+    @staticmethod
+    def denormalize_preferences(normalized_preferences):
+        """
+        Denormalizes preference from simplified to normal structure for saving it in database
+        """
+        denormalized_preferences = {}
+        for app in normalized_preferences.get('apps', []):
+            app_name = app.get('name')
+            app_toggle = app.get('enabled')
+            denormalized_preferences[app_name] = {
+                'enabled': app_toggle,
+                'core_notification_types': normalized_preferences.get('core_notifications', {}).get(app_name, []),
+                'notification_types': {},
+                'non_editable': normalized_preferences.get('non_editable', {}).get(app_name, {}),
+            }
+
+        for preference in normalized_preferences.get('preferences', []):
+            pref_name = preference.get('name')
+            app_name = preference.get('app_name')
+            denormalized_preferences[app_name]['notification_types'][pref_name] = {
+                'web': preference.get('web'),
+                'push': preference.get('push'),
+                'email': preference.get('email'),
+                'info': preference.get('info'),
+            }
+        return denormalized_preferences
+
+    @staticmethod
+    def update_preferences(preferences):
```

Comment:
@muhammadadeeltajamul can we have more details in the method docstring, on how the comparison is calculated.